keps/sig-scheduling/5278-nominated-node-name-for-expectation/README.md

```diff
@@ -0,0 +1,983 @@
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+### The scheduler puts `NominatedNodeName`
+
+After the pod is permitted at `WaitOnPermit`, the scheduler needs to update `NominatedNodeName` with the node that it determines the pod is going to.
+
+Also, in order to set `NominatedNodeName` only when some PreBind plugins work, we need to add a new function (or create a new extension point, if we are concerned about the breaking change to the existing PreBind plugins).
+
+```go
+type PreBindPlugin interface {
+	Plugin
+	// **New Function** (or we can have a separate Plugin interface for this, if we're concerned about a breaking change for custom plugins)
+	// It's called before PreBind, and the plugin is supposed to return Success, Skip, or Error status.
+	// If it returns Skip, it means this PreBind plugin has nothing to do with the pod.
+	// This function should be lightweight, and shouldn't do any actual operation, e.g., creating a volume etc
+	PreBindPreFlight(ctx context.Context, state *CycleState, p *v1.Pod, nodeName string) *Status
+
+	PreBind(ctx context.Context, state *CycleState, p *v1.Pod, nodeName string) *Status
+}
+```
+
+The scheduler would run a new function `PreBindPreFlight()` before `PreBind()` functions, 
+and if all PreBind plugins return Skip status from new functions, we can skip setting `NominatedNodeName`.
+
+This is a similar approach we're doing with PreFilter/PreScore -> Filter/Score. 
+We determine if each plugin is relevant to the pod by Skip status from PreFilter/PreScore, and then determine whether to run Filter/Score function accordingly.
+
+In this way, even if users have some PreBind custom plugins, they can implement `PreBindPreFlight()` appropriately 
+so that the scheduler can wisely skip setting `NominatedNodeName`, taking their custom logic into consideration.
+
+### External components put `NominatedNodeName`
+
+There aren't any restrictions preventing other components from setting NominatedNodeName as of now.
+However, we don't have any validation of how that currently works.
+To support the usecases mentioned above we will adjust the scheduler to do the following:
+- if NominatedNodeName is set, but corresponding Node doesn't exist, kube-scheduler will NOT clear it when the pod is unschedulable [assuming that a node might appear soon]
+- We will rely on the fact that a pod with NominatedNodeName set is resulting in the in-memory reservation for requested resources. 
+Higher-priority pods can ignore it, but pods with equal or lower priority don't have access to these resources. 
+This allows us to prioritize nominated pods when nomination was done by external components. 
```

Comment:
Does that mean scheduler will process pods with NNN set before pods without NNN set as long as they have the same priority?

Reply:
This "prioritize" here is not about the scheduling order. It just says, if pod-A has NNN to node-A, then other pods, except higher priority ones, can steal the space of pod-A on node-A.
I'll change the wording to make it clearer.